### ISSUE TITLE
PDF2 bug? Header and footer missing on last page of ToC

### ISSUE THREAD

**Issue by DJBHollis** [dita-ot/dita-ot#2318]
I'm using DOT 2.1.2 as it is conveniently bundled with a certain XML editor.

I'm using ditamap, not bookmap. `$mirror-page-margins` is false.

For some reason, there is no header or footer on the last page of the ToC. The ToC has three pages. It starts on p.2, and the page with the missing header & footer is p.4.

`layout-masters.xsl` would seem to redirect the last page header and footer to the even page header and footer. I have set up the even page header to be identical to the odd page header, but it doesn't appear on p.4

**Comment by keberlein** [comment]
@DJBHollis The DITA-OT team will need to reproduce this. Can you upload your doc set? Otherwise it might be quite a long time before someone happens to have a sample set at hand which they can use to reproduce the issue.

**Comment by DJBHollis** [comment]
OK, many thanks. Upload to where?

**Comment by keberlein** [comment]
You supposedly can attach files to a GitHub issue. However, I was unable to attach a ZIP today. @robander Any suggestions?

**Comment by raducoravu** [comment]
When you write a comment, at the bottom of the text area there is a label like:

     Attach files by dragging & dropping or by selecting them.

**Comment by xephon2** [comment]
@DJBHollis, do you use an unmodified `pdf2` transformation or do you use an override plugin?

**Comment by DJBHollis** [comment]
Thanks, everyone. For some reason, I assumed there would be a 'bug report' area in the code base. I didn't think to add them to the issue itself!

It's a customisation plug-in.

**Comment by DJBHollis** [comment]
Here are the files ...

[dita_source.zip](https://github.com/dita-ot/dita-ot/files/231535/dita_source.zip)
[com.adaptive.pdf.zip](https://github.com/dita-ot/dita-ot/files/231538/com.adaptive.pdf.zip)
[metadata_manager_user_guide.pdf](https://github.com/dita-ot/dita-ot/files/231539/metadata_manager_user_guide.pdf)

**Comment by xephon2** [comment]
I recommend to output the name of the template directly. Then you can see, whether the correct templates are called. If not, we have to find out why -> Toolkit bug. If yes, you know which templates to fix in your plugin.

```xml
<xsl:template name="insertIndexEvenHeader">
  <fo:static-content flow-name="even-index-header">
    <fo:block>Template called: insertIndexEvenHeader</fo:block>
  </fo:static-content>
</xsl:template>
```

**Comment by DJBHollis** [comment]
The problem is that there is no header on p.4, so I don't see that adding a fo:block to output the name is going to change that.

I don't think the issue is in the headers and footers, I think it is either in the page sequence logic, or how the master page template is called.

**Comment by raducoravu** [comment]
From what I tested this would look like a bug in the RenderX XEP processor.
Your XSLT customization looks good. I also published using Apache FOP, it produced only two table of contents pages and the logo was on each of them.

**Comment by eerohele** [comment]
The issue is that you have `$mirror-page-margins` set to `false()`,  as you said, and in the `static-contents.xsl` file in your plugin, you have:

```xsl
<xsl:template name="insertTocStaticContents">
    <xsl:call-template name="insertTocOddFooter"/>
    <xsl:if test="$mirror-page-margins">
        <xsl:call-template name="insertTocEvenFooter"/>
    </xsl:if>
    <xsl:call-template name="insertTocOddHeader"/>
    <xsl:if test="$mirror-page-margins">
        <xsl:call-template name="insertTocEvenHeader"/>
    </xsl:if>
</xsl:template>
```

In other words, the static contents for even TOC pages is only inserted if `$mirror-page-margins` is `true()`.

If you look at `org.dita.pdf2/cfg/fo/layout-masters.xsl`, you can see that it defines a master for the last TOC page, which uses the same static contents as even TOC pages:

```xsl
<fo:simple-page-master master-name="toc-last" xsl:use-attribute-sets="simple-page-master">
    <fo:region-body xsl:use-attribute-sets="region-body.even"/>
    <fo:region-before region-name="even-toc-header" xsl:use-attribute-sets="region-before"/>
    <fo:region-after region-name="even-toc-footer" xsl:use-attribute-sets="region-after"/>
</fo:simple-page-master>
```

The solution therefore is to change your `static-content.xsl` to:

```xsl
<xsl:template name="insertTocStaticContents">
    <xsl:call-template name="insertTocOddFooter"/>
    <xsl:call-template name="insertTocEvenFooter"/>
    <xsl:call-template name="insertTocOddHeader"/>
    <xsl:call-template name="insertTocEvenHeader"/>
</xsl:template>
```

**Comment by keberlein** [comment]
Hi, @DJBHollis . I'm very glad that you are getting more involved in creating PDF customizations. And @eerohele , many thanks for identifying the problem with David's PDF customization. But a few gentle suggestions:
1. Test and see whether the issue is with the DITA-OT base code, as opposed to your custom plug-in, before opening a DITA-OT issue. You can do this by publishing your content using the OOB PDF2 transformation.
1. If the problem is with your customization, good places to look for help are the dita-users and DITA-OT users lists. Or ping friends with good DITA-OT skills.
1. If the problem seems to be with the DITA-OT code, confirm with the latest stable release of the DITA-OT and then open a issue. Read the guidelines for issue at https://github.com/dita-ot/dita-ot/blob/develop/CONTRIBUTING.md to ensure that your issue meets the guidelines.

**Comment by DJBHollis** [comment]
Hi all,

Thanks for looking into this, and finding a solution.

For me, I consider that there is still an issue surrounding the `$mirror-page-margins` and page sequence logic.

If I set `$mirror-page-margins` to `false()`, then why do I need to create even page headers and footers? I simply should not have to! There should never be a call for an even page.

Or, to put it another way, the logic works for the body section, why not for ToC?

There is another aspect here, that the first page of the ToC is on page 2. If `$mirror-page-margins` were set to `true`, this would be on page 3 because the logic would add a dummy last page to front matter. Typically, this would be blank.

The way the logic is set up, a call to the first page becomes a call to the odd page template. 

So, template wise, in the example content, the ToC is: P2, odd page template; P3, odd page template; P4 even page template.

In other words, there is a bug! IMHO!

Anyway, I now have a workaround. Thanks once again.

**Comment by DJBHollis** [comment]
Just to be clear, the section of code which @eerohele found in the customisation is straight out of the DITA 2.1.2 code. I haven't modified it.

The code that I have modified has 'Adaptive' associated with it. For background, please see: #2319.

**Comment by robander** [comment]
FWIW - I am able to reproduce this with out-of-the-box 2.2.4. I had to add several chapters to David's source map to make it extend to 3 pages. The default header (Adaptive Metadata Manager | Contents | ii) appears on page 2, and on page 3, but on page 4 (third page of TOC) there is no header. If I add enough to get four pages of TOC, the header appears on all four. This does seem like a bug, but I haven't used this part of the code enough to know offhand whether changes will break customizations. 
[metadata_manager_user_guide.pdf](https://github.com/dita-ot/dita-ot/files/232399/metadata_manager_user_guide.pdf)

the map with one chapter duplicated enough to get a third page of TOC.

**Comment by eerohele** [comment]
>If I set `$mirror-page-margins` to `false()`, then why do I need to create even page headers and footers? I simply should not have to! There should never be a call for an even page.

I agree that it would make sense to have `first-toc-header`, `first-toc-footer`, `last-toc-header`, and `last-toc-footer`. The same thing applies for the index and the glossary.

Adding those shouldn't be too difficult. Might make for a good first bug?

**Comment by infotexture** [comment]
See #2112 for a related issue: [Missing attribute sets for Glossary, Index & ToC](https://github.com/dita-ot/dita-ot/issues/2112).

**Comment by SanFanDocu** [comment]
Last page headers of bookmap chapters are missed too, if not followed by a blank page. 
Identified with oxygen 18, standard pdf build (FOP) and build-in 2.x OT or separate 2.2.4 OT.
build-in OT 1.8 works fine.

**Comment by DJBHollis** [comment]
From `layout-masters.xsl`, I put back `<!--blank-or-not-blank="blank"-->` that had been commented out:

```
  <!-- Generate a page sequence master -->
  <xsl:template name="generate-page-sequence-master">
    <xsl:param name="master-name"/>
    <xsl:param name="master-reference"/>
    <xsl:param name="first" select="true()"/>
    <xsl:param name="last" select="true()"/>
    <fo:page-sequence-master master-name="{$master-name}">
      <fo:repeatable-page-master-alternatives>
        <xsl:if test="$first">
          <fo:conditional-page-master-reference master-reference="{$master-reference}-first"
                                                odd-or-even="odd"
                                                page-position="first"/>
        </xsl:if>
        <xsl:if test="$last">
          <fo:conditional-page-master-reference master-reference="{$master-reference}-last"
                                                odd-or-even="even"
                                                page-position="last"/>
                                            <!--blank-or-not-blank="blank"-->
        </xsl:if>
        <xsl:choose>
          <xsl:when test="$mirror-page-margins">
            <fo:conditional-page-master-reference master-reference="{$master-reference}-odd"
                                                  odd-or-even="odd"/>
            <fo:conditional-page-master-reference master-reference="{$master-reference}-even"
                                                  odd-or-even="even"/>
          </xsl:when>
          <xsl:otherwise>
            <fo:conditional-page-master-reference master-reference="{$master-reference}-odd"/>
          </xsl:otherwise>
        </xsl:choose>
      </fo:repeatable-page-master-alternatives>
    </fo:page-sequence-master>
  </xsl:template>
```

It now works as expected.
